### PR TITLE
Added notebook which modifies the preprocessing step for data normalisation

### LINKED ISSUE #28: Normalize training data
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html

### PR DESCRIPTION
**Pull request by eve-b612**
Tested performance with 3 training data normalisation approaches: min max scaling, standard scaling, robust scaling. Results showed no improvement with any of these approaches. 

### Issue Number

* Resolves #28  

### Purpose

*  Modifications to the preprocessing step to test if normalising data would improve model accuracy. 

### Technical Details

* Results did not improve

### Testing Instructions

* If to be used, modify the preprocessing script by writing the additional normalising step (see implementation in cells ## MIN-MAX SCALER, ## STANDARD SCALER, ## ROBUST SCALER)

### Screenshots

* No

## Summary by CodeRabbit

- **Documentation**
	- Added a new section in the README discussing testing of data normalization techniques.
	- Provided insights into the impact of various normalization methods on model accuracy, complementing existing data visualizations.

### CODE REVIEW

**Review by jo-elimu — COMMENTED** [review]
> Tested performance with 3 training data normalisation approaches: min max scaling, standard scaling, robust scaling. Results showed no improvement with any of these approaches.

@eve-b612 Your commit contains 20,457 new lines, so I'm not sure where to look. Would you mind pointing me to the line number where you did the 3 tests?

**Comment by jo-elimu** [comment]
I think I found it now. It's this part, is that correct?:

```json
 {
      "cell_type": "code",
      "source": [
        "## MIN MAX SCALER\n",
        "\n",
        "with open(\"step1_2_preprocess_data.py\", \"r\") as file:\n",
        "    data = file.readlines()\n",
        "\n",
        "# insert min-max Scaler after unnecessary columns are dropped\n",
        "for i, line in enumerate(data):\n",
        "    if \"storybooks_dataframe = storybooks_dataframe[['id', 'reading_level',\" in line:\n",
        "        data.insert(i + 1, \"\"\"\n",
        "from sklearn.preprocessing import MinMaxScaler\n",
        "scaler = MinMaxScaler()\n",
        "storybooks_dataframe[['chapter_count', 'paragraph_count', 'word_count', 'avg_word_length']] = scaler.fit_transform(\n",
        "    storybooks_dataframe[['chapter_count', 'paragraph_count', 'word_count', 'avg_word_length']])\n",
        "print(storybooks_dataframe[['chapter_count', 'paragraph_count', 'word_count', 'avg_word_length']].head())\n",
        "\n",
        "\"\"\")\n",
        "\n",
        "# write to file\n",
        "with open(\"step1_2_preprocess_data.py\", \"w\") as file:\n",
        "    file.writelines(data)\n"
      ],
      "metadata": {
        "id": "wnqmawHJQoGq"
      },
      "execution_count": 26,
      "outputs": []
    }
```

**Comment by eve-b612** [comment]
Yes! So sorry, I'm new to this I should have specified. It's line 5205 for min-max, line 10282 for standard scaler and 15382 for robust scaler. I think it's so long because the notebook outputs are there... I would appreciate any feedback on this approach. Clear outputs? Submit changes in a different format than notebooks? Thanks :)

**Comment by jo-elimu** [comment]
> I would appreciate any feedback on this approach. Clear outputs? Submit changes in a different format than notebooks? Thanks :)

@eve-b612 Well, if you want to be able to easily see how your code changes affected the data further down in the machine learning pipeline, it would probably be easier to add your changes to a new Python file. For example in something like `step1_2_1_normalize_data.py`. Because then you can make your changes, and then run the whole pipeline with `python run_all_steps.py`. Then, if that triggered any changes, you can easily see the diff in Git.

**Comment by eve-b612** [comment]
Ok I see, would you like me to do that? You mentioned that if normalisation did not improve the results then you wouldn't include it, and it in fact did not improve.

**Comment by jo-elimu** [comment]
> Ok I see, would you like me to do that? You mentioned that if normalisation did not improve the results then you wouldn't include it, and it in fact did not improve.

@eve-b612 If the normalization didn't improve the accuracy, then no need to add any more code 👍

But maybe you could add a few words about the experiments you did at the bottom of the `README` just so that we have a record if it?

Thank you for running the tests! 🙂

**Comment by eve-b612** [comment]
No problem! I've just updated the step1_prepare read me file.

**Review by jo-elimu — APPROVED** [review]
